File: javascript/sorting/p160_heapSort.js
/*
    CLRS Section 6.4, p. 160
*/

const { MaxHeap } = require('../data-structures/p157_maxHeap');

const heapsort = (A) => {
    const maxHeap = new MaxHeap(A, A.length);
    maxHeap.buildMaxHeap();
    for(let i = A.length - 1; i >= 1; i--) {
        [A[0], A[i]] = [A[i], A[0]];
        maxHeap.heapSize--;
        maxHeap.maxHeapify(0);
    }
}

module.exports = {
    heapsort
}

File: python/sorting/p160_heapsort.py
'''
    CLRS Section 6.4, p. 160
'''

from data_structures.p157_max_heap import MaxHeap

def heapsort(A):
    max_heap = MaxHeap(A, len(A))
    max_heap.build_max_heap()
    for i in range(len(A)-1, 0, -1):
        A[0], A[i] = A[i], A[0]
        max_heap.heap_size -= 1
        max_heap.max_heapify(0)
